include/seqan3/io/alignment_file/format_sam.hpp

```diff
@@ -793,49 +790,52 @@ inline void format_sam::write_alignment_record(stream_type & stream,
     // ---------------------------------------------------------------------
     // Writing the Record
     // ---------------------------------------------------------------------
-    std::cpp20::ostreambuf_iterator stream_it{stream};
-    char const separator{'\t'};
 
-    write_range(stream_it, std::forward<id_type>(id));
+    detail::fast_ostreambuf_iterator stream_it{*stream.rdbuf()};
+    constexpr char separator{'\t'};
 
-    stream << separator;
+    write_range_or_asterisk(stream_it, id);
+    *stream_it = separator;
 
-    stream << static_cast<uint16_t>(flag) << separator;
+    stream_it.write_number(static_cast<uint16_t>(flag));
+    *stream_it = separator;
 
     if constexpr (!detail::decays_to_ignore_v<ref_id_type>)
     {
         if constexpr (std::integral<std::remove_reference_t<ref_id_type>>)
         {
-            write_range(stream_it, (header.ref_ids())[ref_id]);
+            write_range_or_asterisk(stream_it, (header.ref_ids())[ref_id]);
         }
         else if constexpr (detail::is_type_specialisation_of_v<std::remove_reference_t<ref_id_type>, std::optional>)
         {
             if (ref_id.has_value())
-                write_range(stream_it, (header.ref_ids())[ref_id.value()]);
+                write_range_or_asterisk(stream_it, (header.ref_ids())[ref_id.value()]);
             else
-                stream << '*';
+                *stream_it = '*';
         }
         else
         {
-            write_range(stream_it, std::forward<ref_id_type>(ref_id));
+            write_range_or_asterisk(stream_it, ref_id);
         }
     }
     else
     {
-        stream << '*';
+        *stream_it = '*';
     }
 
-    stream << separator;
+    *stream_it = separator;
 
     // SAM is 1 based, 0 indicates unmapped read if optional is not set
-    stream << (ref_offset.value_or(-1) + 1) << separator;
+    stream_it.write_number(ref_offset.value_or(-1) + 1);
+    *stream_it = separator;
 
-    stream << static_cast<unsigned>(mapq) << separator;
+    stream_it.write_number(static_cast<unsigned>(mapq));
+    *stream_it = separator;
 
     if (!std::ranges::empty(cigar_vector))
     {
-        for (auto & c : cigar_vector)
-            stream << c.to_string(); // returns a small_vector instead of char so write_range doesn't work
+        for (auto & c : cigar_vector) //TODO THIS IS PROBABLY TERRIBLE PERFORMANCE_WISE
```

Comment:
Did you see/measure any performance impact or does it still need to be confirmed?
Or, do you already have any idea what could be done differently?

Reply:
> Did you see/measure any performance impact or does it still need to be confirmed?

It's just an intuition. The string returned is stack-allocated, so it's not terribly bad, but it still means we are unpacking every single element into one string and than copying that string into another.

> Or, do you already have any idea what could be done differently?

A private helper function could remedy this (basically doing what `.to_string()`does but writing directly into the target).

I have quite a few other things/ideas regarding the SAM format handling, I just didn't want us to forget this.